### PR TITLE
OCSADV-426: Added checks for executed observations.

### PR DESCRIPTION
**Pull request by sraaphorst**
Previously, BAGS was selecting guide stars for observations marked as "executed".
We now leave these observations alone.

@swalker2m, as per your suggestion due to the computation-heavy nature of calls to `ObservationStatus.computeFor`, I tried to push this call as far as possible into the BAGS process, but I am not sure if this is the best place to put it (i.e. after AGS is called but before results are applied), and the calls to AGS are redundant in this case.

I would greatly appreciate any feedback or suggestions as to if you think there is a preferable place to put this, as the calls to AGS are unnecessary with the results thrown away as things currently stand.

### CODE REVIEW

File: bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala [hunk 1]
```diff
@@ -50,10 +51,12 @@ object BagsManager {
         Try { GemsGuideStarWorker.findGuideStars(obsCtx) } match {
           case Success(ggs) =>
             LOG.info(s"BAGS GeMS lookup for observation=${observation.getObservationID} successful. ${resultString(ggs.asScalaOpt)}")
-            Swing.onEDT {
-              muteObservation(observation)
-              GemsGuideStarWorker.applyResults(TpeContext(observation), ggs, true)
-              unmuteObservation(observation)
+            if (ObservationStatus.computeFor(observation) != ObservationStatus.OBSERVED) {
```
Comment (cquiroz) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala]: Both code fragments look very similar, could they be refactored?

**Comment by sraaphorst** [comment]
I made the refactor as suggested by @cquiroz (which I think is a very good idea).
I'm not entirely happy with the way that the code looks, as there is ugliness from having to mix the Gemini `Option` (which is used by `GemsGuideStarWorker`) with the Scala `Option` (which is used by `GuideStarWorker`). If anyone has any suggestions to get around this without mucking things up further, I would very much be happy to hear! @swalker2m or @tpolecat ?

File: bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala [hunk 2]
```diff
@@ -39,54 +42,33 @@ object BagsManager {
...
+              Swing.onEDT {
+                muteObservation(observation)
+                worker(TpeContext(observation), selOpt)
+                unmuteObservation(observation)
+              }
             }
             taskComplete(this, success = true)
+
           case Failure(ex) =>
             LOG.warning(s"BAGS lookup for observation=${observation.getObservationID} failed. Exception=$ex")
```
Comment (tpolecat) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala]: Not part of this PR, but would you mind including the stacktrace via `Log.log(Level.WARNING, ..., ex)`?

**Comment by tpolecat** [comment]
Yeah the `Option` thing is a pain. Seems fine.

File: bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala (hunk 2 above)
Comment (sraaphorst) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala]: I don't mind at all. My only concern is that this log message is output if BAGS is attempting to run and the network is out, which could get very verbose.

Perhaps I should match on the exception and output a separate message if that is the case. The only problem I see with this is that a `GenericError` is what is reported if the catalog host is unreachable, and from what I can tell, this may be issued in other circumstances.

For now, for a `GenericError`, I will simply output the message, and for any other error type, I will output the stack trace.

Comment (cquiroz) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala]: Now that you mention the catalog calls, there could be a concern on affecting the scalability of the catalog servers if we have all the OTs doing queries in the background. I don't know if that is a real concern though
What if the user drags a target, would it do a query at the end or on multiple steps?

Comment (sraaphorst) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala]: It will only do a query at the end of the drag, as I believe that the `TargetEnvironment` is only changed when the drag is complete.

**Comment by sraaphorst** [comment]
Okay, all concerns have been addressed, so I am going to merge this one.
Added the stack trace for exceptions in logging (trying to avoid network outages) as per the request by @tpolecat
